File: 프로그래머스/lv3/최원준/자물쇠와 열쇠.py
'''
1. 아이디어 :
    구현문제
2. 시간복잡도 :
    O( n**2 * m**2 )
3. 자료구조 :
    배열

'''

def solution(key, lock):
    def rotate_key():
        n = len(key)
        for j in range(n // 2):
            for i in range(n - 1 - 2 * j):
                key[j][j + i], key[j + i][n - 1 - j], key[n - 1 - j][n - 1 - j - i], key[n - 1 - j - i][j] = key[n - 1 - j - i][j], key[j][j + i], key[j + i][n - 1 - j], key[n - 1 - j][n - 1 - j - i]


    def check(x, y, total):
        n = len(lock)
        m = len(key)
        for row in range(m):
            for col in range(m):
                if 0 <= x + row < n and 0 <= y + col < n:
                    if key[row][col] == 1 and lock[x + row][y + col] == 1:
                        return False
                    if key[row][col] == 1 and lock[x + row][y + col] == 0:
                        total -= 1
        return total == 0

    holes = 0
    for row in range(len(lock)):
        for col in range(len(lock)):
            if lock[row][col] == 0:
                holes += 1

    for row in range(-len(key)+1, len(lock)):
        for col in range(-len(key)+1, len(lock)):
            for i in range(4):
                if check(row, col, holes):
                    return True
                rotate_key()
    return False

File: 프로그래머스/lv3/지준혁/자물쇠와 열쇠.cpp
#include <string>
#include <vector>

using namespace std;

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    bool answer = true;
    
    check_unlock(); // key에 있는 값을 lock으로 옮겼을 떄 lock이 다 1로 되는지
    
    int len = 
    for (int roate = 0; rotate < 4; ++rotate) {
        rotate_key(); // 키 90도 회전 시키기
                      // key 상하좌우로 이동 가능한 경우의 수 4^20? 시간초과//
    }
    
    return answer;
}
